### PR TITLE
feat(witness): use block executor to execute block inside debug_execution_witness

### PR DESCRIPTION
**Pull request by 0x00101010**
## Description

The code does the following:
1. allow passing `BlockExecutorProvider` to `RpcModuleBuilder` and eventually passed down to `DebugApi` initialization
2. rework `debug_execution_witness` API to use block_executor to execute blocks.

### CODE REVIEW

File: crates/ethereum/evm/src/execute.rs [hunk 1]
```diff
@@ -381,6 +384,39 @@ where
...
+                for (slot, value) in &account.storage {
+                    let slot = B256::from(*slot);
+                    let hashed_slot = keccak256(slot);
+                    storage.storage.insert(hashed_slot, *value);
+
+                    state_preimages.insert(hashed_slot, alloy_rlp::encode(slot).into());
+                }
+            }
+        }
+        println!("state_preimages: {:?}", state_preimages);
```
Comment (Rjected) [inline crates/ethereum/evm/src/execute.rs:413]: I don't think we need to do this inside the executor, this should still occur in `debug_executionWitness` rpc

File: crates/rpc/rpc/src/debug.rs [hunk 2]
```diff
@@ -388,7 +396,7 @@ where
                                 Ok(frame.into())
                             })
                             .await?;
-                        return Ok(frame)
+                        return Ok(frame);
```
Comment (Rjected) [inline crates/rpc/rpc/src/debug.rs:399]:
```suggestion
                        return Ok(frame)
```
should also configure the editor to not insert semicolons

**Review by Rjected — CHANGES_REQUESTED** [review]
It would be good to start with integrating the block executor into the `debug_executionWitness` rpc, and then debugging. Is the assert failing due to changes in this PR or does it also occur on main? If it occurs on main, definitely submit an issue and we can focus on it there

File: crates/ethereum/evm/src/execute.rs (hunk 1 above)
Comment (0x00101010) [reply crates/ethereum/evm/src/execute.rs:413]: This feels like a natural place to do it?

The plan is to add a `include_preimages` bool in the input and only optionally generate it if configured

**Comment by 0x00101010** [comment]
@Rjected here's the issue: https://github.com/paradigmxyz/reth/issues/10756

File: crates/ethereum/evm/src/execute.rs (hunk 1 above)
Comment (0x00101010) [reply crates/ethereum/evm/src/execute.rs:413]: Or else, we would need to ask block_executor to return `State<DB>` back for debug_execution_witness to be able to access other vars in `State<DB>` like `cache`,

Comment (Rjected) [reply crates/ethereum/evm/src/execute.rs:413]: Ah, I missed that the current `BlockExecutionOutput` does not have all the information required to include preimages. IMO it would be nice to have separation of concerns, between block execution and fetching preimages though.

One idea is to change the `BlockExecutionOutput` to include the `CacheState`, that way anyone who has a `BlockExecutionOutput` can generate preimages, wdyt?

Comment (0x00101010) [reply crates/ethereum/evm/src/execute.rs:413]: yeah, that works!

Comment (0x00101010) [reply crates/ethereum/evm/src/execute.rs]: @Rjected does this look good to you? https://github.com/bluealloy/revm/pull/1756, curious how do we handle revm release cycle here?

Asking because we can directly take the cache out without the convenience method defined in revm, it is a nice to have to keep things consistent.

File: crates/ethereum/evm/src/execute.rs [hunk 3]
```diff
@@ -378,10 +375,14 @@ where
         let EthExecuteOutput { receipts, requests, gas_used } =
             self.execute_without_verification(block, total_difficulty)?;
 
-        // NOTE: we need to merge keep the reverts for the bundle retention
-        self.state.merge_transitions(BundleRetention::Reverts);
-
-        Ok(BlockExecutionOutput { state: self.state.take_bundle(), receipts, requests, gas_used })
+        let cache = core::mem::take(&mut self.state.cache);
```
Comment (0x00101010) [inline crates/ethereum/evm/src/execute.rs:378]: replace with self.state.take_cache() when https://github.com/bluealloy/revm/pull/1756 is merged

File: crates/optimism/evm/src/execute.rs [hunk 4]
```diff
@@ -352,8 +352,10 @@ where
         // NOTE: we need to merge keep the reverts for the bundle retention
         self.state.merge_transitions(BundleRetention::Reverts);
 
+        let cache = core::mem::take(&mut self.state.cache);
```
Comment (0x00101010) [inline crates/optimism/evm/src/execute.rs:355]: replace with self.state.take_cache() when https://github.com/bluealloy/revm/pull/1756 is merged

File: crates/rpc/rpc/src/debug.rs (hunk 2 above)
Comment (0x00101010) [reply crates/rpc/rpc/src/debug.rs]: Done with `make fmt`

File: crates/ethereum/evm/src/execute.rs [hunk 5]
```diff
@@ -381,7 +381,13 @@ where
         // NOTE: we need to merge keep the reverts for the bundle retention
         self.state.merge_transitions(BundleRetention::Reverts);
 
-        Ok(BlockExecutionOutput { state: self.state.take_bundle(), receipts, requests, gas_used })
+        Ok(BlockExecutionOutput {
+            state: self.state.take_bundle(),
+            cache: core::mem::take(&mut self.state.cache),
```
Comment (0x00101010) [inline crates/ethereum/evm/src/execute.rs:386]: extract to a function in revm: https://github.com/bluealloy/revm/pull/1756

**Comment by rkrasiuk** [comment]
@mattsse thoughts on adding executor factory to rpc?

**Comment by mattsse** [comment]
> @mattsse thoughts on adding executor factory to rpc?

that makes sense, I assume this could be useful and we could extend this functionality and use this as the entrypoint to abstract execution in rpc in general

File: crates/evm/src/test_utils.rs [hunk 6]
```diff
@@ -60,6 +61,10 @@ impl<DB> Executor<DB> for MockExecutorProvider {
             gas_used: 0,
         })
     }
+
+    fn state_ref(&self) -> &State<DB> {
+        todo!()
+    }
```
Comment (mattsse) [inline crates/evm/src/test_utils.rs:67]: this now bakes in the state into the executor interface, I don't want to do this

File: crates/evm/src/test_utils.rs [hunk 7]
```diff
@@ -50,7 +51,7 @@ impl<DB> Executor<DB> for MockExecutorProvider {
     type Output = BlockExecutionOutput<Receipt>;
     type Error = BlockExecutionError;
 
-    fn execute(self, _: Self::Input<'_>) -> Result<Self::Output, Self::Error> {
+    fn execute(&mut self, _: Self::Input<'_>) -> Result<Self::Output, Self::Error> {
```
Comment (mattsse) [inline crates/evm/src/test_utils.rs:54]: why is this necessary?

File: crates/rpc/rpc/src/debug.rs [hunk 8]
```diff
@@ -569,110 +575,80 @@ where
...
+                //             let slot = B256::from(slot);
+                //             let hashed_slot = keccak256(slot);
+                //             storage.storage.insert(hashed_slot, value);
+
+                //             if include_preimages {
+                //                 state_preimages.insert(hashed_slot,
+                // alloy_rlp::encode(slot).into());             }
+                //         }
+                //     }
+                // }
```
Comment (emhane) [inline crates/rpc/rpc/src/debug.rs:647]: probably should remove this before marking pr ready for review, else mark pr as draft

**Review by emhane — COMMENTED** [review]
would be nice with a test for `DebugApi::debug_execution_witness`, however possibly running hive tests for this branch would be enough before merging? wdyt @Rjected ?
